Kahlan 3.0.0

jails opened the pull request:
# Kahlan 3.0.0

I took the opportunity for this new Kahlan milestone to BC break the `toReceivedNext` matcher and also to clean up the codebase to be able to add a couple of new features.

## BC break:
* The `'params'` option of `Stub::create()` has been renamed to `'args'`.
* `toReceivedNext` has been removed in flavor of `->ordered` attribute to be more close to Rspec way.
* `Stub::on()` is now "deprecated" in flavor of `allow()`.
* `Monkey::patch()` is now "deprecated" in flavor of `allow()`.
* `Stub::create()` has been renamed to `Double::instance()` (to be kind of "consistent" with instance_double() in RSpec).
* `Stub::classname()` has been renamed to `Double::classname()` (to be kind of "consistent" with class_double() in RSpec).
* `before()` and `after()` have been renamed to `beforeAll()` and `afterAll()` (BC Break).
* `Args` has been renamed `CommandLine` (i.e. `$this->args()` become `$this->commandLine()` in kahlan-config.php).

```php
it("expects \$foo to receive message() followed by foo()", function() {
    $foo = new Foo();
    expect($foo)->toReceive('message')->ordered;
    expect($foo)->toReceive('foo')->ordered;
    $foo->message();
    $foo->foo();
});
```

## New DSL for Stubbing/Monkey patching:

Since the RSpec philosophy is based on the fact specs should be easily readable. I introduced a unified DSL around Stubbing/Monkey patching (similar to `allow()` in RSpec):

```php

// Monkey patch `PDO` and stub chained methods.
allow('PDO')->toReceive('prepare', 'fetchAll')->andReturn([['name' => 'bob']]);
allow('PDO')->toReceive('prepare', 'fetchAll')->andReturnUsing(function() {
    return [['name' => 'bob']];
});

// Monkey patch `PDO` with a specific class.
allow('PDO')->toBe('My\Alternative\PDO'); 

// Monkey patch `DateTime` with a specific instance.
allow('DateTime')->toBe(new DateTime('@123'));

// Monkey patch `PDO` with a generic stub instance.
allow('PDO')->toBeOK(); 

// Monkey patch a function and set a return value
allow('time')->toBeCalled()->andReturn(123);
allow('time')->toBeCalled()->andReturnUsing(function() { return 123; });

// Monkey patch `getmxrr()` with an empty closure.
allow('getmxrr')->toBeOK(); 
```

Using the previous Syntax:

```php

// Monkey patch `PDO` and stub chained methods.
$pdo = Stub::classname();
Monkey::patch('PDO', $pdo);
Stub::on($pdo)->method('prepare', 'fetchAll')->andReturn([['name' => 'bob']]);

// Monkey patch `PDO` with a specific class.
Monkey::patch('PDO', 'My\Alternative\PDO');

// Monkey patch `DateTime` with a specific instance.
Monkey::patch('DateTime', new DateTime('@123')); // But was unsupported in Kahlan < 3.0

// Monkey patch `PDO` with a generic stub instance.
Monkey::patch('PDO', Stub::create());

// Monkey patch a function and set a return value
Monkey::patch('time', function() { return 123; });

// Monkey patch `getmxrr()` with an empty closure.
Monkey::patch('time', function() {});
```

## New features:

Kahlan now offer a way to deal with long chain of method calls: 

```php
$instance->a()->c()->c()->d();
```
In practice methods chains is more considered as code smells because it tends to violate the Law of Demeter. Even though not all code smells indicate real problems so I ported the RSpec `receive_message_chain` up to Kahlan through the following notation:

```php
it("expects called chain to be called", function() {
    $foo = new Foo();
    allow($foo)->toReceive('a', 'b', 'c')->andReturn('something');
    expect($foo)->toReceive('a', 'b', 'c'); // you can check that the call actually happened but it will be a redundant test most of the time
    expect($foo->a()->b()->c())->toBe('something');
});
```

This pattern can also be used with core classes when passed as a fully namespaced class name. The class will be automatically monkey patched.

Example:
```php
it('should patch PDO', function() {
    allow('PDO')->toReceive('prepare', 'fetchAll')->andReturn([['name' => 'bob']]);

    $user = new User();
    expect($user->all())->toBe([
        ['name' => 'bob']
    ]);
});
```

Where the `User` class is something like:

```php
<?php
use PDO;

class User
{
    protected $_db;

    public function __construct()
    {
        $this->_db = new PDO('mysql:dbname=testdb;host=localhost', 'root','');
    }

    public function all()
    {
        $stmt = $this->db->prepare('SELECT * FROM users');
        $stmt->execute();
        return $stmt->fetchAll();
    }
}
```

## expect()->toReceived()/toBeCalled():

Expect only spy for method/function calls. So stubbing & spying need to be done separately and can be done independently.

```php
it("patches a core function", function() {
    $foo = new Foo();
    allow('time')->toBeCalled()->andReturn(245026800);
    expect('time')->toBeCalled();
    expect($foo->time())->toBe(245026800);
});
```

jails commented:
@m1ome can you play with this dev branch a bit and give me your feedback ?

jails commented:
@jasonmccreary since you were interested to get involved in Kahlan your feedback is welcome ;-)

m1ome commented:
@jails ok i will try on weekends.

jasonmccreary commented:
Thanks. I do want to get more involved. Unfortunately Laravel Shift has taken all my extra time with the recent Laravel 5.3 release. Definitely will take a look though as I'm giving a TDD in PHP workshop in November and want to demo Kahlan.

jails commented:
I added the `allow()` statement to build stubs ala RSpec and make stub creation more readable, what do you think about the DSL ?

m1ome commented:
@jails for me looks pretty good, try to run on weekends, don't see any big difference. For me at leasst 👍

warrenseymour commented:
Awesome PR! My thoughts follow. Please note that I'm not fully versed with the name of the original RSpec, so some of my comments may be ill-founded:

* `andReturnUsing` seems like it would read better as `andCall`, like so:

```php
allow('PDO')->toReceive('prepare->fetchAll')->andCall(function() {
    return [['name' => 'bob']];
});
```

* Clarification on `allow($foo)->toReceive('bar')->andReturn('baz');`

Would it be correct to say that such a spec would fail if $foo->bar() was never called? If so, that's great as it tidies up a lot of patterns where I have to explicitly say `Stub::on...` and `expect()->toReceive`, but it should be made more apparent that this new construct integrates stubbing **and** assertions into one thing.

* Chained call parameter assertions

Love the new approach to allow assertions again chained calls. Is it possible to make assertions against the parameters of each call in the chain? For example:

```php
$db->find('widgets)
  ->where(['name' => 'Bottle Opener'])
  ->order(['id' => 'desc'])
  ->limit(10)
```

```
php
expect($db)
  ->toReceive('find->where->order->limit')
  ->where([
    'find' => 'widgets',
    'where' => ['name' => 'Bottle Opener'],
    'order' => ['id => 'desc'],
    'limit' => 10
  ]);
```

jails commented:
* Renaming `andReturnUsing` to `andCall`:

Good call !  By the way does it still ok for functions ? `allow('time')->toBeCalled()->andCall(...);` ? What about `allow('time')->toBeCalled()->andRun(...);`. I'm not a native english speaker so I'm sometimes confused by naming.

* Clarification on `allow($foo)->toReceive('bar')->andReturn('baz');`

You made a good point. I merged both concept through `expect()` first but it tended to be ambiguous. Indeed sometimes `expect()` acted as a simple "spy" but was also able to set some stubbing when used with `andReturn()`. So in `expect($something)->toReceive('method')->andReturn(true);` for example, it didn't mean that we are expecting `$something->method()` to be called and return `true` but expecting `$something->method()` to be called and then just bail out the method to return `true`.

So I thought I would be better to separate both task to avoid this ambiguity. That's why I introduced the RSpec's `allow()` function. Anyhow it tends to has a fuzzy meaning. Maybe it would be better to rename it to `stub()` or `patch()` to avoid confusion.

So with the chain example:

```php
$db->find('widgets')
  ->where(['name' => 'Bottle Opener'])
  ->order(['id' => 'desc'])
  ->limit(10)
```

The following syntax won't work:
```php
expect($db)
  ->toReceive('find->where->order->limit')
  ->where([
    'find' => 'widgets',
    'where' => ['name' => 'Bottle Opener'],
    'order' => ['id => 'desc'],
    'limit' => 10
  ]);
```
Unfortunately it's not possible to spy chains with current Kahlan's JIT implementation. Actual poincuts can be changed by using closures but it will add an overhead penalty on code execution time and will make the backtrace different that the one you would get on executing real code.

So this call will throw the following error: "Kahlan can't Spy chained methods on real PHP code, you need to Stub the chain first."

But you'll be able to write:
```php
patch($pb)->toReceive('find->where->order->limit')->andReturn([]);
expect($db)
  ->toReceive('find->where->order->limit')
  ->where([
    'find' => 'widgets',
    'where' => ['name' => 'Bottle Opener'],
    'order' => ['id => 'desc'],
    'limit' => 10
  ]);
```

or simply:
```php
patch($db)
  ->toReceive('find->where->order->limit')
  ->where([
    'find' => 'widgets',
    'where' => ['name' => 'Bottle Opener'],
    'order' => ['id => 'desc'],
    'limit' => 10
  ])
  ->andReturn([]);
```
You're not explicitly checking that methods has been called but by setting a return value based on constraints, you should be able to check that the call has been made if the returned result has been correctly build.
Alternatly with can also introduce the following syntax:

```php
$patch = patch($db)
  ->toReceive('find->where->order->limit')
  ->where([
    'find' => 'widgets',
    'where' => ['name' => 'Bottle Opener'],
    'order' => ['id => 'desc'],
    'limit' => 10
  ])
  ->andReturn([]);
expect($patch)->toBeCalled();
```
Does that make sense ?

warrenseymour commented:
* What about allow('time')->toBeCalled()->andRun(...);

You're right - I think `andRun()...` reads better than `andCall()`. The main reason I dislike `andReturnUsing()` is because you may be stubbing smelly code that has side effects and the return value may not be particularly meaningful.

The restrictions on chained call assertions are reasonable, it was more of a request for a 'bonus feature' on top of new API than any kind of hard requirement from me personally.

Thanks for all the hard work on Kahlan!

jasonmccreary commented:
I really like the RSpec direction. This was going to be one of my main contributions when I first looked at Kahlan.

I know I'm a little late, but what do you think about some of the following direction changes.

- What if there was no differentiation between a stub/spy/patch? Just simplify it to `allow()`. Is there a technical reason these need to be differentiated? Otherwise, it would greatly reduce the learning curve for new users adopting Kahlan. Personally this is something I like about other Spec based frameworks (RSPec, Jasmine, Cedar, etc).
- What do you think about the method chaining to take variadic arguments? For example, `allow($foo)->toReceive('a', 'b', 'c')` instead of `allow($foo)->toReceive('a->b->c')`. I worry the latter will lead to many typos since its a string. At least the argument separation breaks it up.

As far as other items, I like `andRun()` instead of repeating `toBeCalled()` + `andCall()`. However, I'd rather change the way native functions are setup to keep the `andCall()`. I'll try to think of an example later tonight.

jasonmccreary commented:
Thought about it, what about allowing shorthands like `toReturn` and `toCall`.

For example:

```php
allow('time')->toReturn(5);
allow('time')->toCall(function() {
    return 5;
});
```

It reads well and remove the repetitive `toBeCalled`. I mean, when would you not have to type `toBeCalled()` in these cases? Never, right? So just condense it to KISS.

jails commented:
Cool thanks for all the feedback guys !

You made a valid point @warrenseymour and this feature would definitely move things forward. And it shouldn't be that hard to implement.

Happy you get a chance to look at this new Kahlan's version @jasonmccreary !

> What if there was no differentiation between a stub/spy/patch? Just simplify it to allow()

At this point that's exactly what `allow()` is trying to do. Merging the both `Stub::on()` and `Monkey::patch()` into an unified API. I used the RSpec naming at this point but `allow()` seems to be a awkward naming. So I'm wondering if `patch()` wouldn't be a better naming that `allow()`. I'm always up to not reinvent the wheel but in this case I find the `allow()` naming a bit confusing.

To sum up changing the behavior of something will be done using `patch()` like so.
```php
patch('Model')->toReceive('::load')->with(123)->andReturn(['id' => 123, 'name' => 'bob']);
```
Note that under the hood everything will be either monkey patched or stubbed to make it works.

And the second syntax (which didn't change from the current version) is to listen for messages:
```php
expect('Model')->toReceive('::load')->with(123);
```
This syntax just "listen/spy" messages but **didn't change the code behavior** and the code will run as usual.

> What do you think about the method chaining to take variadic arguments? For example, allow($foo)->toReceive('a', 'b', 'c') instead of allow($foo)->toReceive('a->b->c').

Good point. Having all the `->` to make the string to looks like a fluent syntax is probably not necessary.

> Thought about it, what about allowing shorthands like toReturn and toCall.

The challenge behind all of this was to keep the API for both `patch()` and `expect()` the most close as possible to avoid confusion.

Let's take the following as example:
```php
// Makes the rand function to return 5 when called with (0, 10) as arguments.
patch('rand')->toBeCalled()->with(0, 10)->andReturn(5);

// Expect that rand has been called with (0, 10).
expect('rand')->toBeCalled()->with(0, 10);
```

If `toBeCalled()` seems a bit superfluous at some point for the stubbing part, it tends to be a no brainer. Whatever you do, if it's related to a function it's `toBeCalled()` and if it's a class it's `toReceive()`.

Btw `toBeCalled()` has been introduced in this new version, so feel free to propose whatever you want if you think about a better naming which can be shared for both `patch()` and `expect()`.

jasonmccreary commented:
Makes sense to keep the APIs similar. However, my vote is for `allow()` over `patch()`. I appreciate *patch* might be a more technical term for what's actually happening, but it is not a common testing term. As such, it would be at the bottom of my choices. I'd even rather see `stub()` than `patch()`. 😉

jails commented:
Right, `stub()` can be a good compromise :+1: ! It's a common testing term and have a strong meaning.

jails commented:
And what about `make()` ?

jails commented:
Thanks guys for all the feedback.
The PR is ready to be merged and will be tagged anytime soon.

I updated the documentation so you can [find changes here](http://kahlan.readthedocs.io/en/dev/)

To sum up:
* I finally kept the `allow()` statement to not disturb people with some RSpec experience.
* Method chaining make use of variadic arguments instead of '->'
* It's possible to apply arguments requirements on a chain using `where()`.
* `Stub::on()` has been deprecated in flavor of `allow()`
* `Monkey::patch()` has been deprecated in flavor of `allow()`
* The `'params'` option of `Stub::create()` has been renamed to `'args'`.
* `Stub::create()` has been renamed to `Double:instance()`
* `Stub::classname()` has been renamed to `Double:classname()`
* `toReceivedNext` has been removed in flavor of `->ordered` attribute to be more close to Rspec way.

The 3.0.0 now take advantage of the (PHP Throwable interface)[http://php.net/manual/en/class.throwable.php]. So no more fatal error when running PHP >= 7.0. Everything will shut down nicely with a fancy console report while PHP >= 5.4 is still supported.

It's now possible to mark specs as pending by adding no expectation inside or mak them as excluded when using the `xit`, `xcontext`, `xdescribe` syntax.

The reporting has also been slightly modified to reflect all changes related to this new version.

And this 3.0.0 officially support PHP 7.1

jasonmccreary commented:
Awesome!

A few last thoughts...

Is it possible to also *focus* tests with `fit`, `fcontext`, `fdescribe`?

How many people are running Kahlan currently? If there's enough, I'm glad to maybe make some kind of *shifter* as I have for [upgrading PHP](https://php-shift.com) and [upgrading Laravel](https://laravelshift.com).

jails commented:
`fit`, `fcontext` and `fdescribe` has been included since 1.0 so I think we are good ;-) !
It's a good question. Currently there's approx. 100 install per day on average so it's limited.

jasonmccreary commented:
So I'd like to still contribute to the 3.0 release somehow. Maybe I can update the English documentation?

jails commented:
Would be fantastic, since English in not my mother tongue the documentation is in a kind of broken english and probably incomprehensible ! The `docs` directory in the `dev` branch contains all changes related to 3.0, and any help on that part will be more than welcome.

jasonmccreary commented:
Great. I'll focus on that side of it for now until I learn the internals a bit more to contribute.

jasonmccreary commented:
What do you think about dropping `before` and `after`?

Mainly for the fact that they are confusing on when to use, especially for new testers, and often come with side effects. As such, their use is ultimately discouraged. Kahlan could avoid this issue by simply not including these constructs.

At the least, I suggest changing their name to `beforeAll` and `afterAll` to add clarity and symmetry with their `*Each` counterparts.

jails commented:
Makes sense. Agreed that "before/after" should be used with caution. But I don't think it worth a remove. They are the only existing workaround to speed up test suites with expensive setup and teardown.
Btw you made a good call, the 3.0 is the perfect time to rename them `beforeAll` and `afterAll` :+1:

jasonmccreary commented:
Fair enough. We should revisit this though as other performance opportunities reveal themselves.

I'm going to give the docs a first pass on Sunday. Look for the initial PR late evening.

jails commented:
Perfect :100: !

jasonmccreary commented:
Had some time this afternoon and took a quick run through the docs. A few more ideas to run by you:

- What do you think about `toBeOfType` instead of `toBeA`. While I like the simplicity of `toBeA` is the improper indefinite article for `int`, `array`, or `object` (e.g. `toBeA('int')`). `toBeOfType` or `toBeTypeOf` in similar with `toBeInstanceOf` and avoids using indefinite article. Or, since there aren't that many data types in PHP, just create explicit `toBeAnInteger`, `toBeAFloat`, etc.
- To the point of matchers, is there anything explicitly for `true`/`false`. I see *truthy*/*falsy*. But most matcher grammars include a strict `toBeTrue`/`toBeFalse`.

Also, how do you pronounce *Kahlan*? I've been pronouncing it like the name *Colin*. Maybe wanted to throw a note in the docs as it will help avoid confusion and promote people talking about *Kahlan*.

jails commented:
Great ! It's always interesting to have a second opinion.

* There's actually `toBeA()` and `toBeAn()` which are both aliases [on the same matcher](https://github.com/crysalead/kahlan/blob/master/src/Cli/Kahlan.php#L313-L314). I chose this option mainly because RSpec are using `be_a` `be_an` so I didn't want to change people's mind. That's why I would prefer keep them for "interoperability" but I'm fine to add new alias if necessary.

* Explicit matcher for true/false are `toBe(true)`/`toBe(false)`. Didn't find `toBeTruthy()`/`toBeFalsy()` naming that great either but it was the Jasmine way, so for the same reason I preferred to keep them as is. Personally I'm writing `toBe(null)` instead of `toBeNull()`. However `toBeNull()` has been included in Kahlan so all Jasmine matchers can be used.

It's a good question, I think I pronounce it the french way. In fact the name come from [a television series](https://en.wikipedia.org/wiki/Legend_of_the_Seeker). The heroine is a Confessor and her power lies in the revelation of the truth. Which is what tests are for, reveling the truth ;-) !
I'm not an expert but imo [@warrenseymour pronunciation](https://vimeo.com/116949820) is correct.

jasonmccreary commented:
I agree with the `toBeA` and `toBeAn`. I should have asked if there was the *An* version.

I'd still advocate for a strict `toBeTrue` and `toBeFalse`. Given the weak typing of PHP, I would expect `toBeTruthy` and `toBeFalsy` to behave more *loosely*. I'm glad to add these to get familiar with the core.

I'll check out the reference and be sure to include it in the docs. Thanks.

jails commented:
Sure, since `toBeNull()` already exists I see no reason not to add `toBeTrue()` &  `toBeFalse()` at least for consistency :+1:

angelxmoreno commented:
love it

jasonmccreary commented:
Are you comfortable adopting the PSR-2 style guide? I'd like to add this as part of the `Contributing` section of the documentation and setup StyleCI to run on PRs if so.

jails commented:
I'm totally fine with adopting a coding style. But this organization (of a handful of people) really get on my nerves. Would prefer to not give them further credit.
However I progressively changed Kahlan's coding style on point which didn't matter like (tab/space, lowercase/CamelCase namespace, where `{` should be opened, etc.)
So I think we are PSR-2 compatible somehow.

PSR-2 aside, would be great to add the following rules if possible:
* `private` visibility is forbidden
* `final` visibility is forbidden
* type hinting is forbidden
* `protected` methods/variables must be prefixed with a single underscore.

jasonmccreary commented:
We can resist PSR-2 later as those items would require some custom rules. Just curious if it was close.

As far as the rest, what do you think about putting Kahlan into its own organization. So `kahlan/kahlan`? It would be easier to break out the matchers and stubs into projects later if we wanted (like Rspec does).

jails commented:
I wanted to do it a couple of years ago but looked like the `kahlan` organization already existed. And since I didn't know how to do so without breaking packagist, I gave up.

jasonmccreary commented:
You can rename a project with Packagist. It'll still require people to update their dependencies, but *3.0* would be a perfect opportunity to do so.

If `kahlan/kahlan` isn't available, you could put it under your own account if you like. It's your choice.

I'm almost done with a first pass of the documentation. Will open a PR tonight (tomorrow morning for you).

jails commented:
I'll try to claim the `kahlan` account first since it looks like a dead account according to github [policy](https://help.github.com/articles/name-squatting-policy/).

Awesome, looking forward seeing your PR.

jails commented:
Fantastic github released the username for us https://github.com/kahlan!

m1ome commented:
@jails awesome!

samsonasik commented:
@jails I created learn-kahlan repository and provide its kahlan v3 support PR https://github.com/sitrunlab/learn-kahlan/pull/2 , the not works part is on replacing `Stub::on` with `allow`:
```
PHP Fatal error:  Call to undefined method Kahlan\Allow::method()
 in /home/travis/build/sitrunlab/learn-kahlan/spec/FooSpec.php on line 42
```
this is the latest travis build: https://travis-ci.org/sitrunlab/learn-kahlan/jobs/161287341#L189

samsonasik commented:
oh, sorry, wrong repo

samsonasik commented:
pull `dev` and run `bin/kahlan --config=kahlan-config.travis.php --clover=clover.xml` on my local got the following error:
```bash
Dir
  ::tempnam()
    ✘ it uses the system temp directory by default
      expect->toMatch() failed in `./spec/Suite/Dir/Dir.spec.php` line 312
      
      It expect actual to match expected.
      
      actual:
        (string) "/private/var/folders/q6/4grzj1mj50g2fflkfl94yjp00000gp/T/specwxMJvR"
      expected:
        (string) "~^/var/folders/q6/4grzj1mj50g2fflkfl94yjp00000gp/T/spec~"
      


  Skipped specifications: 3
  ./spec/Suite/Scope.spec.php, line 150
  ./spec/Suite/Scope.spec.php, line 157
  ./spec/Suite/Scope.spec.php, line 175

Expectations   : 1897 Executed
Specifications : 0 Pending, 0 Excluded, 3 Skipped

Passed 899 of 900 FAIL (FAILURE: 1) in 22.866 seconds (using 58Mo)
```
My environment:
```bash
Apples-MBP:kahlan samsonasik$ php -v
PHP 7.0.11 (cli) (built: Sep 22 2016 09:11:47) ( NTS )
Copyright (c) 1997-2016 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2016 Zend Technologies
    with Zend OPcache v7.0.11, Copyright (c) 1999-2016, by Zend Technologies
    with Xdebug v2.4.1, Copyright (c) 2002-2016, by Derick Rethans

Apples-MBP:kahlan samsonasik$ echo MAJOR_MAC_VERSION=$(sw_vers -productVersion | awk -F '.' '{print $1 "." $2}')
MAJOR_MAC_VERSION=10.11

Apples-MBP:kahlan samsonasik$ uname -v
Darwin Kernel Version 15.0.0: Wed Aug 26 16:57:32 PDT 2015; root:xnu-3247.1.106~1/RELEASE_X86_64
```

jails commented:
@samsonasik thanks for the head up, I updated the spec so now it should pass. I also moved documentation inside its dedicated repo (https://github.com/kahlan/docs) including your fixes .

samsonasik commented:
it works now 👍

samsonasik commented:
rebased against master. conflicts resolved.

samsonasik commented:
is `composer.json`'s description need to be updated, currently:
```
Behavior-Driven Development (BDD) library.
```
kahlan baseline itself is:
```
The PHP Test Framework for Freedom, Truth and Justice.
```
We use both `framework` and `library` term for it ?

jails commented:
Yup :+1:
Better to have `The PHP Test Framework for Freedom, Truth and Justice.` everywhere !

samsonasik commented:
xdebug:master is now successfully installed when using php 7.1 on travis, the strange things are it shows error: `Xdebug MUST be loaded as a Zend extension in Unknown on line 0` and some errors on specs, @jails could you check ?

samsonasik commented on .travis.yml:
```diff
@@ -21,7 +21,7 @@ before_script:
   - composer install --no-interaction
   - if [[ $TRAVIS_PHP_VERSION =~ ^hhvm ]]; then echo 'hhvm.php7.all = 1' >> /etc/hhvm/php.ini; fi
-  - if [[ $TRAVIS_PHP_VERSION =~ ^7 ]]; then pecl install xdebug; fi
+  - if [[ $TRAVIS_PHP_VERSION =~ ^7 ]]; then git clone git://github.com/xdebug/xdebug.git && cd xdebug && phpize && ./configure --enable-xdebug  && make && make install && echo "extension = xdebug.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini && cd ..; fi
```
because `pecl install xdebug` was return error

jails commented:
@samsonasik imo use `echo "zend_extension =  xdebug.so"` instead of `echo "extension = xdebug.so"`

samsonasik commented:
the error is now only in specs:
```
Lcov
  ::export()
    ✖ it exports the coverage of a file with no extra end line
      an uncaught exception has been thrown in `src/Reporter/Coverage/Exporter/Lcov.php` line 83
      message:`Kahlan\PhpErrorException` Code(0) with message "`E_WARNING` A non-numeric value encountered"
```

samsonasik commented:
@jails it fixed now, please verify my latest commit: https://github.com/kahlan/kahlan/pull/155/commits/4fc990f87e972906dac1e3e3b3d51c3ff2391a4d

jails commented:
Good catch 👍 !

jasonmccreary commented:
I'm not seeing my changes in the new `kahlan/docs` repo… Shall I just create a new `3.0` branch there? Also, I see you're using [Couscous](http://couscous.io) now. Since this gives us more flexibility with the theme, are you comfortable if I match the [Jasmine docs](http://jasmine.github.io/2.5/introduction.html)? When done right, as Jasmine has, I love the side by side descriptions and code.

jails commented:
Weird, maybe you didn't pushed all your changes @jasonmccreary ? Just verified and the seven commits you made above are applied in `kahlan/docs`.

Yeah you can create the branch you want, `3.0`, `dev` whatever. And for the "Jasmine docs" theme it's a good idea. It's not the best format for mobile phone but it's not a big deal for such documentation imo.

LostKobrakai commented:
Would be nice if the each text/example group could stack on top of each other for mobile. I've used the kahlan docs from time to time in a separate window next to the editor, which on a laptop will not be to much bigger than for a mobile phone.

jails commented:
Fair enough, do you think it's something doable @jasonmccreary ?

jasonmccreary commented:
I thought you had merged/pushed them. No worries since we are going in the direction of the Jasmine style docs anyway. It shouldn't be a problem stacking the sections on mobile. I'll ensure that's part of the final design.

I'll try to have an initial version of the new Jasmine direction up by the end of the week. Future comments will be on the `kahlan/docs` repo.

jails commented:
@jasonmccreary awesome !
I'm tagging this and let's jump on `kahlan/docs`.
